craft_parts/executor/part_handler.py

```diff
@@ -292,14 +396,29 @@ def _run_overlay(
         else:
             contents = StepContents()
 
+        partitions_contents: dict[str, MigrationContents] = {
+            p: MigrationContents(files=c.files, directories=c.dirs)
+            for p, c in contents.partitions_contents.items()
+            if p is not DEFAULT_PARTITION
+        }
+
         layer_hash = self._compute_layer_hash(all_parts=False)
         layer_hash.save(self._part)
 
+        default_files: set[str] = set()
+        default_directories: set[str] = set()
+
+        default_contents = contents.partitions_contents.get(DEFAULT_PARTITION)
+        if default_contents:
+            default_files = default_contents.files
+            default_directories = default_contents.dirs
+
         return states.OverlayState(
             part_properties=self._part_properties,
             project_options=step_info.project_options,
-            files=contents.files,
-            directories=contents.dirs,
+            partitions_contents=partitions_contents,
+            files=default_files,
+            directories=default_directories,
```

Comment:
Maybe this could become a bit more readable if `default_files` and `default_directories` are not declared at all, and `default_contents` defaults to an empty instance of `MigationContents` (set in line 411)?